openwisp_controller/pki/tests/test_api.py

```diff
@@ -0,0 +1,153 @@
+            r = self.client.get(path)
+        self.assertEqual(r.status_code, 200)
+
+    def test_ca_delete_api(self):
+        ca1 = self._create_ca(name='ca1', organization=self._get_org())
+        path = reverse('pki_api:ca_detail', args=[ca1.pk])
+        r = self.client.delete(path)
+        self.assertEqual(r.status_code, 204)
+        self.assertEqual(Ca.objects.count(), 0)
+
+    def test_cert_post_api(self):
+        path = reverse('pki_api:cert_list')
+        data = self._get_cert_data.copy()
+        data['ca'] = self._create_ca().pk
+        with self.assertNumQueries(8):
+            respone = self.client.post(path, data, content_type='application/json')
+        self.assertEqual(respone.status_code, 201)
+        self.assertEqual(Cert.objects.count(), 1)
+
+    def test_cert_list_api(self):
+        self._create_cert(name='cert1')
+        path = reverse('pki_api:cert_list')
+        with self.assertNumQueries(4):
+            response = self.client.get(path)
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(Cert.objects.count(), 1)
+
+    def test_cert_detail_api(self):
+        cert1 = self._create_cert(name='cert1')
+        path = reverse('pki_api:cert_detail', args=[cert1.pk])
+        response = self.client.get(path)
+        with self.assertNumQueries(4):
+            response = self.client.get(path)
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.data['id'], cert1.pk)
+
+    def test_cert_put_api(self):
+        cert1 = self._create_cert(name='cert1')
+        org2 = self._create_org()
+        path = reverse('pki_api:cert_detail', args=[cert1.pk])
+        data = {
+            'name': 'cert1-change',
+            'organization': org2.pk,
+            'notes': 'new-notes',
+        }
+        with self.assertNumQueries(8):
+            response = self.client.put(path, data, content_type='application/json')
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.data['name'], 'cert1-change')
+        self.assertEqual(response.data['organization'], org2.pk)
+        self.assertEqual(response.data['notes'], 'new-notes')
+
+    def test_cert_patch_api(self):
+        cert1 = self._create_cert(name='cert1')
+        path = reverse('pki_api:cert_detail', args=[cert1.pk])
+        data = {'name': 'cert1-change'}
+        with self.assertNumQueries(7):
+            response = self.client.patch(path, data, content_type='application/json')
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.data['name'], 'cert1-change')
+
+    def test_cert_delete_api(self):
+        cert1 = self._create_cert(name='cert1')
+        path = reverse('pki_api:cert_detail', args=[cert1.pk])
+        with self.assertNumQueries(8):
+            response = self.client.delete(path)
+        self.assertEqual(response.status_code, 204)
+        self.assertEqual(Cert.objects.count(), 0)
```

Comment:
please add `assertNumQueries` to each test to ensure we're keeping queries in check

Reply:
Done adding. :+1: